File: src/frontend/parser.c

#include "parser.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <shared/helper/mem.h>

struct parser *parser_create(struct lexer *lexer) {
    struct parser *parser = xmalloc(sizeof(struct parser));
    parser->lexer = lexer;
    parser->lookahead = NULL;
    return parser;
}

void parser_destroy(struct parser *parser) {
    if (parser->lookahead != NULL) {
        lexer_destroy_token(parser->lookahead);
    }
    xfree(parser);
}

static int parser_read_lookahead(struct parser *parser) {
    struct lexer_token *token;
    if (lexer_read_next_token(parser->lexer, &token)) {
        return -1;
    }
    parser->lookahead = token;
    return 0;
}

static int parser_match(struct parser *parser, enum lexer_token_type type, struct lexer_token **token) {
    if (parser->lookahead->type == type) {
        if (token == NULL) {
            lexer_destroy_token(parser->lookahead);
        } else {
            *token = parser->lookahead;
        }
        parser->lookahead = NULL;
        return parser_read_lookahead(parser);
    } else {
        lexer_print_location(stderr, parser->lookahead->location);
        fprintf(stderr, ": Expected %s but got %s\n", lexer_token_type_name(type), lexer_token_type_name(parser->lookahead->type));
        return -1;
    }
}

/*
 * blocks               -> ( block )*
 * block                -> 'block' identifier '{' declarations behaviour_statements '}'
 * declarations         -> ( declaration )*
 * declaration          -> type identifier_list ';'
 * identifier_list      -> identifier ( ',' identifier )*
 * type                 -> type_specifier [ '[' number ']' ]
 * type_specifier       -> 'in' | 'out' | 'block' identifier
 * behaviour_statements -> ( behaviour_statement )*
 * behaviour_statement  -> behaviour_identifier '=' expression ';'
 * behaviour_identifier -> identifier [ '.' identifier ] [ subscript ]
 * subscript            -> '[' number [ ':' number ] ']'
 */

/*
 * Operators:
 *
 * Unary:
 *   Bitwise NOT: ~
 * Bitwise AND: &
 * Bitwise XOR: ^
 * Bitwise OR: |
 */

/*
 * expr -> bit_or_expr
 * bit_or_expr -> bit_xor_expr ( '|' bit_xor_expr )*
 * bit_xor_expr -> bit_and_expr ( '^' bit_and_expr )*
 * bit_and_expr -> unary_expr ( '&' unary_expr )*
 * unary_expr -> '~' unary_expr | primary_expr
 * primary_expr -> '(' expr ')' | behaviour_identifier | number
 */

static int parse_identifier(struct parser *parser, struct ast_node **node) {
    int ret;

    struct lexer_token *token;

    ret = parser_match(parser, TOKEN_IDENTIFIER, &token);
    if (ret) {
        return ret;
    }

    *node = ast_create_node(AST_IDENTIFIER);
    (*node)->data.identifier = xstrdup(token->value.identifier);

    lexer_destroy_token(token);

    return 0;
}

static int parse_number(struct parser *parser, struct ast_node **node) {
    int ret;

    struct lexer_token *token;

    ret = parser_match(parser, TOKEN_NUMBER, &token);
    if (ret) {
        return ret;
    }

    *node = ast_create_node(AST_NUMBER);
    (*node)->data.number = token->value.number;

    lexer_destroy_token(token);

    return 0;
}

static int parse_expression(struct parser *parser, struct ast_node **node);
static int parse_behaviour_identifier(struct parser *parser, struct ast_node **node);

static int parse_primary_expression(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *expression = NULL;
    struct ast_node *behaviour_identifier = NULL;
    struct ast_node *number = NULL;

    switch (parser->lookahead->type) {
        case TOKEN_PARENTHESIS_LEFT:
            ret = parser_match(parser, TOKEN_PARENTHESIS_LEFT, NULL);
            if (ret) {
                goto err;
            }

            ret = parse_expression(parser, &expression);
            if (ret) {
                goto err;
            }

            ret = parser_match(parser, TOKEN_PARENTHESIS_RIGHT, NULL);
            if (ret) {
                goto err;
            }

            *node = expression;

            break;
        case TOKEN_IDENTIFIER:
            ret = parse_behaviour_identifier(parser, &behaviour_identifier);
            if (ret) {
                goto err;
            }

            *node = behaviour_identifier;

            break;
        case TOKEN_NUMBER:
            ret = parse_number(parser, &number);
            if (ret) {
                goto err;
            }

            *node = number;

            break;
        default:
            lexer_print_location(stderr, parser->lookahead->location);
            fprintf(stderr, ": Expected '(', identifier or number\n");
            ret = -1;
            goto err;
    }

    return 0;

err:
    ast_destroy_node(expression);
    ast_destroy_node(behaviour_identifier);
    ast_destroy_node(number);
    return ret;
}

static int parse_unary_expression(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *unary_expression = NULL;
    struct ast_node *primary_expression = NULL;

    switch (parser->lookahead->type) {
        case TOKEN_NOT:
            ret = parser_match(parser, TOKEN_NOT, NULL);
            if (ret) {
                goto err;
            }

            ret = parse_unary_expression(parser, &unary_expression);
            if (ret) {
                goto err;
            }

            *node = ast_create_node(AST_UNARY_EXPRESSION);
            (*node)->data.op = AST_OP_NOT;
            ast_add_child_node(*node, unary_expression);

            break;
        default:
            ret = parse_primary_expression(parser, &primary_expression);
            if (ret) {
                goto err;
            }

            *node = primary_expression;
    }

    return 0;

err:
    ast_destroy_node(unary_expression);
    ast_destroy_node(primary_expression);
    return ret;
}

static int parse_bitwise_and_expression(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *left = NULL;
    struct ast_node *right = NULL;

    ret = parse_unary_expression(parser, &left);
    if (ret) {
        goto err;
    }

    while (parser->lookahead->type == TOKEN_AND) {
        ret = parser_match(parser, TOKEN_AND, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_unary_expression(parser, &right);
        if (ret) {
            goto err;
        }

        struct ast_node *new_left = ast_create_node(AST_BINARY_EXPRESSION);
        new_left->data.op = AST_OP_AND;
        ast_add_child_node(new_left, left);
        ast_add_child_node(new_left, right);

        left = new_left;
        right = NULL;
    }

    *node = left;

    return 0;

err:
    ast_destroy_node(left);
    ast_destroy_node(right);
    return ret;
}

static int parse_bitwise_xor_expression(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *left = NULL;
    struct ast_node *right = NULL;

    ret = parse_bitwise_and_expression(parser, &left);
    if (ret) {
        goto err;
    }

    while (parser->lookahead->type == TOKEN_XOR) {
        ret = parser_match(parser, TOKEN_XOR, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_bitwise_and_expression(parser, &right);
        if (ret) {
            goto err;
        }

        struct ast_node *new_left = ast_create_node(AST_BINARY_EXPRESSION);
        new_left->data.op = AST_OP_XOR;
        ast_add_child_node(new_left, left);
        ast_add_child_node(new_left, right);

        left = new_left;
        right = NULL;
    }

    *node = left;

    return 0;

err:
    ast_destroy_node(left);
    ast_destroy_node(right);
    return ret;
}

static int parse_bitwise_or_expression(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *left = NULL;
    struct ast_node *right = NULL;

    ret = parse_bitwise_xor_expression(parser, &left);
    if (ret) {
        goto err;
    }

    while (parser->lookahead->type == TOKEN_OR) {
        ret = parser_match(parser, TOKEN_OR, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_bitwise_xor_expression(parser, &right);
        if (ret) {
            goto err;
        }

        struct ast_node *new_left = ast_create_node(AST_BINARY_EXPRESSION);
        new_left->data.op = AST_OP_OR;
        ast_add_child_node(new_left, left);
        ast_add_child_node(new_left, right);

        left = new_left;
        right = NULL;
    }

    *node = left;

    return 0;

err:
    ast_destroy_node(left);
    ast_destroy_node(right);
    return ret;
}

static int parse_expression(struct parser *parser, struct ast_node **node) {
    return parse_bitwise_or_expression(parser, node);
}

static int parse_subscript(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *number_start = NULL;
    struct ast_node *number_end = NULL;

    ret = parser_match(parser, TOKEN_BRACKET_LEFT, NULL);
    if (ret) {
        goto err;
    }

    ret = parse_number(parser, &number_start);
    if (ret) {
        goto err;
    }

    // Optional end index
    if (parser->lookahead->type == TOKEN_COLON) {
        ret = parser_match(parser, TOKEN_COLON, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_number(parser, &number_end);
        if (ret) {
            goto err;
        }
    }

    ret = parser_match(parser, TOKEN_BRACKET_RIGHT, NULL);
    if (ret) {
        goto err;
    }

    *node = ast_create_node(AST_SUBSCRIPT);
    ast_add_child_node(*node, number_start);
    ast_add_child_node(*node, number_end);

    return 0;

err:
    ast_destroy_node(number_start);
    ast_destroy_node(number_end);
    return ret;
}

static int parse_behaviour_identifier(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *identifier = NULL;
    struct ast_node *property_identifier = NULL;
    struct ast_node *subscript = NULL;

    ret = parse_identifier(parser, &identifier);
    if (ret) {
        goto err;
    }

    // Optional property access
    if (parser->lookahead->type == TOKEN_DOT) {
        ret = parser_match(parser, TOKEN_DOT, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_identifier(parser, &property_identifier);
        if (ret) {
            goto err;
        }
    }

    // Optional subscript
    if (parser->lookahead->type == TOKEN_BRACKET_LEFT) {
        ret = parse_subscript(parser, &subscript);
        if (ret) {
            goto err;
        }
    }

    *node = ast_create_node(AST_BEHAVIOUR_IDENTIFIER);
    ast_add_child_node(*node, identifier);
    ast_add_child_node(*node, property_identifier);
    ast_add_child_node(*node, subscript);

    return 0;

err:
    ast_destroy_node(identifier);
    ast_destroy_node(property_identifier);
    ast_destroy_node(subscript);
    return ret;
}

static int parse_behaviour_statement(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *behaviour_identifier = NULL;
    struct ast_node *expression = NULL;

    ret = parse_behaviour_identifier(parser, &behaviour_identifier);
    if (ret) {
        goto err;
    }

    ret = parser_match(parser, TOKEN_EQUALS, NULL);
    if (ret) {
        goto err;
    }

    ret = parse_expression(parser, &expression);
    if (ret) {
        goto err;
    }

    ret = parser_match(parser, TOKEN_SEMICOLON, NULL);
    if (ret) {
        goto err;
    }

    *node = ast_create_node(AST_BEHAVIOUR_STATEMENT);
    ast_add_child_node(*node, behaviour_identifier);
    ast_add_child_node(*node, expression);

    return 0;

err:
    ast_destroy_node(behaviour_identifier);
    ast_destroy_node(expression);
    return ret;
}

static int parse_behaviour_statements(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *behaviour_statements = NULL;
    struct ast_node *behaviour_statement = NULL;

    while (parser->lookahead->type == TOKEN_IDENTIFIER) {
        ret = parse_behaviour_statement(parser, &behaviour_statement);
        if (ret) {
            goto err;
        }

        struct ast_node *new_behaviour_statements;

        if (behaviour_statements == NULL) {
            new_behaviour_statements = behaviour_statement;
        } else {
            new_behaviour_statements = ast_create_node(AST_BEHAVIOUR_STATEMENTS);
            ast_add_child_node(new_behaviour_statements, behaviour_statements);
            ast_add_child_node(new_behaviour_statements, behaviour_statement);
        }

        behaviour_statements = new_behaviour_statements;
        behaviour_statement = NULL;
    }

    *node = behaviour_statements;

    return 0;

err:
    ast_destroy_node(behaviour_statements);
    ast_destroy_node(behaviour_statement);
    return ret;
}

static int parse_identifier_list(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *left = NULL;
    struct ast_node *right = NULL;

    ret = parse_identifier(parser, &left);
    if (ret) {
        goto err;
    }

    while (parser->lookahead->type == TOKEN_COMMA) {
        ret = parser_match(parser, TOKEN_COMMA, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_identifier(parser, &right);
        if (ret) {
            goto err;
        }

        struct ast_node *new_left = ast_create_node(AST_IDENTIFIER_LIST);
        ast_add_child_node(new_left, left);
        ast_add_child_node(new_left, right);

        left = new_left;
        right = NULL;
    }

    *node = left;

    return 0;

err:
    ast_destroy_node(left);
    ast_destroy_node(right);
    return ret;
}

static int parse_type_specifier(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *identifier = NULL;

    switch (parser->lookahead->type) {
        case TOKEN_KEYWORD_IN:
            ret = parser_match(parser, TOKEN_KEYWORD_IN, NULL);
            if (ret) {
                goto err;
            }

            *node = ast_create_node(AST_TYPE_SPECIFIER_IN);

            break;
        case TOKEN_KEYWORD_OUT:
            ret = parser_match(parser, TOKEN_KEYWORD_OUT, NULL);
            if (ret) {
                goto err;
            }

            *node = ast_create_node(AST_TYPE_SPECIFIER_OUT);

            break;
        case TOKEN_KEYWORD_BLOCK:
            ret = parser_match(parser, TOKEN_KEYWORD_BLOCK, NULL);
            if (ret) {
                goto err;
            }

            ret = parse_identifier(parser, &identifier);
            if (ret) {
                goto err;
            }

            *node = ast_create_node(AST_TYPE_SPECIFIER_BLOCK);
            ast_add_child_node(*node, identifier);

            break;
        default:
            lexer_print_location(stderr, parser->lookahead->location);
            fprintf(stderr, ": Expected type\n");
            ret = -1;
            goto err;
    }

    return 0;

err:
    ast_destroy_node(identifier);
    return ret;
}

static int parse_type(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *type_specifier = NULL;
    struct ast_node *number = NULL;

    ret = parse_type_specifier(parser, &type_specifier);
    if (ret) {
        goto err;
    }

    // Optional type width
    if (parser->lookahead->type == TOKEN_BRACKET_LEFT) {
        ret = parser_match(parser, TOKEN_BRACKET_LEFT, NULL);
        if (ret) {
            goto err;
        }

        ret = parse_number(parser, &number);
        if (ret) {
            goto err;
        }

        ret = parser_match(parser, TOKEN_BRACKET_RIGHT, NULL);
        if (ret) {
            goto err;
        }
    }

    *node = ast_create_node(AST_TYPE);
    ast_add_child_node(*node, type_specifier);
    ast_add_child_node(*node, number);

    return 0;

err:
    ast_destroy_node(type_specifier);
    ast_destroy_node(number);
    return ret;
}

static int parse_declaration(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *type = NULL;
    struct ast_node *identifier_list = NULL;

    ret = parse_type(parser, &type);
    if (ret) {
        goto err;
    }

    ret = parse_identifier_list(parser, &identifier_list);
    if (ret) {
        goto err;
    }

    ret = parser_match(parser, TOKEN_SEMICOLON, NULL);
    if (ret) {
        goto err;
    }

    *node = ast_create_node(AST_DECLARATION);
    ast_add_child_node(*node, type);
    ast_add_child_node(*node, identifier_list);

    return 0;

err:
    ast_destroy_node(type);
    ast_destroy_node(identifier_list);
    return ret;
}

static int parse_declarations(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *declarations = NULL;
    struct ast_node *declaration = NULL;

    while (parser->lookahead->type == TOKEN_KEYWORD_IN
        || parser->lookahead->type == TOKEN_KEYWORD_OUT
        || parser->lookahead->type == TOKEN_KEYWORD_BLOCK
        ) {
        ret = parse_declaration(parser, &declaration);
        if (ret) {
            goto err;
        }

        struct ast_node *new_declarations;

        if (declarations == NULL) {
            new_declarations = declaration;
        } else {
            new_declarations = ast_create_node(AST_DECLARATIONS);
            ast_add_child_node(new_declarations, declarations);
            ast_add_child_node(new_declarations, declaration);
        }

        declarations = new_declarations;
        declaration = NULL;
    }

    *node = declarations;

    return 0;

err:
    ast_destroy_node(declarations);
    ast_destroy_node(declaration);
    return ret;
}

static int parse_block(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *identifier = NULL;
    struct ast_node *declarations = NULL;
    struct ast_node *behaviour_statements = NULL;

    ret = parser_match(parser, TOKEN_KEYWORD_BLOCK, NULL);
    if (ret) {
        goto err;
    }

    ret = parse_identifier(parser, &identifier);
    if (ret) {
        goto err;
    }

    ret = parser_match(parser, TOKEN_BRACE_LEFT, NULL);
    if (ret) {
        goto err;
    }

    ret = parse_declarations(parser, &declarations);
    if (ret) {
        goto err;
    }

    ret = parse_behaviour_statements(parser, &behaviour_statements);
    if (ret) {
        goto err;
    }

    ret = parser_match(parser, TOKEN_BRACE_RIGHT, NULL);
    if (ret) {
        goto err;
    }

    *node = ast_create_node(AST_BLOCK);
    ast_add_child_node(*node, identifier);
    ast_add_child_node(*node, declarations);
    ast_add_child_node(*node, behaviour_statements);

    return 0;

err:
    ast_destroy_node(identifier);
    ast_destroy_node(declarations);
    ast_destroy_node(behaviour_statements);
    return ret;
}

static int parse_blocks(struct parser *parser, struct ast_node **node) {
    int ret;

    struct ast_node *blocks = NULL;
    struct ast_node *block = NULL;

    while (parser->lookahead->type == TOKEN_KEYWORD_BLOCK) {
        ret = parse_block(parser, &block);
        if (ret) {
            goto err;
        }

        struct ast_node *new_blocks;

        if (blocks == NULL) {
            new_blocks = block;
        } else {
            new_blocks = ast_create_node(AST_BLOCKS);
            ast_add_child_node(new_blocks, blocks);
            ast_add_child_node(new_blocks, block);
        }

        blocks = new_blocks;
        block = NULL;
    }

    *node = blocks;

    return 0;

err:
    ast_destroy_node(blocks);
    ast_destroy_node(block);
    return ret;
}

int parser_parse(struct parser *parser, struct ast_node **root_out) {
    int ret;

    clock_t start = clock();

    ret = parser_read_lookahead(parser);
    if (ret) {
        return ret;
    }

    struct ast_node *root;

    ret = parse_blocks(parser, &root);
    if (ret) {
        return ret;
    }

    if (parser->lookahead->type != TOKEN_EOF) {
        lexer_print_location(stderr, parser->lookahead->location);
        fprintf(stderr, ": Unexpected %s, expected block\n", lexer_token_type_name(parser->lookahead->type));
        ast_destroy_node(root);
        return -1;
    }

    *root_out = root;

    clock_t end = clock();
    double parse_time = (double) (end - start) / CLOCKS_PER_SEC;
    fprintf(stderr, "parsing took %f seconds\n", parse_time);

    return 0;
}

File: src/frontend/parser.rs

use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

use derive_more::Display;

use crate::frontend::lexer;
use super::lexer::{ILexer, TokenKind, Token, Location, TokenData};
use super::ast::*;
use crate::shared::error;

#[derive(Debug, Display)]
pub enum ErrorKind {
    #[display(fmt = "expected {} but got {} at {}", _1, _2, _0)]
    UnexpectedToken(Location, TokenKind, TokenKind),
    #[display(fmt = "expected {} but got {} at {}", _1, _2, _0)]
    UnexpectedToken2(Location, String, TokenKind),
    #[display(fmt = "failed to read next token")]
    Lexer,
}

pub type Error = error::Error<ErrorKind>;

impl From<lexer::Error> for Error {
    fn from(err: lexer::Error) -> Self {
        Error::with_source(ErrorKind::Lexer, err)
    }
}

pub struct Parser<L: ILexer> {
    lexer: L,
    lookahead: Box<Token>
}

impl<L: ILexer> Parser<L> {
    pub fn new(mut lexer: L) -> Result<Parser<L>, Error> {
        // Read first lookahead token
        let lookahead = lexer.get_token()?;

        Ok(Parser {
            lexer,
            lookahead: Box::new(lookahead)
        })
    }

    fn match_token(&mut self, expected_kind: TokenKind) -> Result<Box<Token>, Error> {
        if self.lookahead.kind == expected_kind {
            // Read next lookahead token
            let mut token = Box::new(self.lexer.get_token()?);
            // Swap the new and the old lookahead tokens
            std::mem::swap(&mut self.lookahead, &mut token);
            // `token` now holds the old lookahead token which is the token we just matched against
            Ok(token)
        } else {
            Err(ErrorKind::UnexpectedToken(self.lookahead.location, expected_kind, self.lookahead.kind).into())
        }
    }

    fn parse_identifier(&mut self) -> Result<Box<IdentifierNode>, Error> {
        let identifier = self.match_token(TokenKind::Identifier)?;

        if let TokenData::Identifier(name) = identifier.data {
            Ok(Box::new(IdentifierNode {
                value: name
            }))
        } else {
            panic!("Token `kind` and `data` do not match up");
        }
    }

    fn parse_number(&mut self) -> Result<Box<NumberNode>, Error> {
        let number = self.match_token(TokenKind::Number)?;

        if let TokenData::Number { value, width } = number.data {
            Ok(Box::new(NumberNode {
                value,
                width
            }))
        } else {
            panic!("Token `kind` and `data` do not match up");
        }
    }

    fn parse_string(&mut self) -> Result<Box<StringNode>, Error> {
        let identifier = self.match_token(TokenKind::String)?;

        if let TokenData::String(value) = identifier.data {
            Ok(Box::new(StringNode {
                value
            }))
        } else {
            panic!("Token `kind` and `data` do not match up");
        }
    }

    /// Parser grammar (EBNF):
    ///
    /// ```ebnf
    /// root                 = includes blocks ;
    /// includes             = { include } ;
    /// include              = 'include', string ;
    /// blocks               = { block } ;
    /// block                = 'block', identifier, '{', declarations, behaviour_statements, '}' ;
    /// declarations         = { declaration } ;
    /// declaration          = type, identifier_list, ';' ;
    /// identifier_list      = identifier, { ',', identifier } ;
    /// type                 = type_specifier, [ '[', number, ']' ] ;
    /// type_specifier       = 'in' | 'out' | 'block', identifier ;
    /// behaviour_statements = { behaviour_statement } ;
    /// behaviour_statement  = behaviour_identifier, '=', expr, ';' ;
    /// behaviour_identifier = identifier, [ '.', identifier ], [ subscript ] ;
    /// subscript            = '[', number, [ ':', number ], ']' ;
    ///
    ///
    /// Operators (highest precedence level first):
    ///
    /// Unary:
    ///   Bitwise NOT: ~
    /// Bitwise AND: &
    /// Bitwise XOR: ^
    /// Bitwise OR: |
    ///
    ///
    /// expr         = bit_or_expr ;
    /// bit_or_expr  = bit_xor_expr, { '|', bit_xor_expr } ;
    /// bit_xor_expr = bit_and_expr, { '^', bit_and_expr } ;
    /// bit_and_expr = unary_expr, { '&', unary_expr } ;
    /// unary_expr   = '~', unary_expr | primary_expr ;
    /// primary_expr = '(', expr, ')' | behaviour_identifier | number ;
    /// ```

    pub fn parse(&mut self) -> Result<Box<RootNode>, Error> {
        let includes = self.parse_includes()?;

        let blocks = self.parse_blocks()?;

        // Make sure there are no tokens left over
        self.match_token(TokenKind::EndOfFile)?;

        Ok(Box::new(RootNode {
            includes,
            blocks,
            blocks_map: HashMap::new()
        }))
    }

    fn parse_includes(&mut self) -> Result<Vec<Box<IncludeNode>>, Error> {
        let mut includes = Vec::new();

        while self.lookahead.kind == TokenKind::IncludeKeyword {
            let include = self.parse_include()?;

            includes.push(include);
        }

        Ok(includes)
    }

    fn parse_include(&mut self) -> Result<Box<IncludeNode>, Error> {
        self.match_token(TokenKind::IncludeKeyword)?;

        let name = self.parse_string()?;

        Ok(Box::new(IncludeNode {
            name,
            full_path: None,
        }))
    }

    fn parse_blocks<'a>(&mut self) -> Result<Vec<Rc<RefCell<BlockNode>>>, Error> {
        let mut blocks = Vec::<Rc<RefCell<BlockNode>>>::new();

        while self.lookahead.kind == TokenKind::BlockKeyword {
            let block = self.parse_block()?;

            blocks.push(block);
        }

        Ok(blocks)
    }

    fn parse_block<'a>(&mut self) -> Result<Rc<RefCell<BlockNode>>, Error> {
        self.match_token(TokenKind::BlockKeyword)?;

        let name = self.parse_identifier()?;

        self.match_token(TokenKind::LeftBrace)?;

        let declarations = self.parse_declarations()?;

        let behaviour_statements = self.parse_behaviour_statements()?;

        self.match_token(TokenKind::RightBrace)?;

        Ok(Rc::new(RefCell::new(BlockNode {
            name,
            declarations,
            behaviour_statements,
            symbol_table: None,
            intermediate_block: None,
        })))
    }

    fn parse_declarations(&mut self) -> Result<Vec<Box<DeclarationNode>>, Error> {
        let mut declarations = Vec::<Box<DeclarationNode>>::new();

        while self.lookahead.kind == TokenKind::InKeyword
            || self.lookahead.kind == TokenKind::OutKeyword
            || self.lookahead.kind == TokenKind::BlockKeyword {

            let declaration = self.parse_declaration()?;

            declarations.push(declaration);
        }

        Ok(declarations)
    }

    fn parse_declaration(&mut self) -> Result<Box<DeclarationNode>, Error> {
        let typ = self.parse_type()?;

        let names = self.parse_identifier_list()?;

        self.match_token(TokenKind::Semicolon)?;

        Ok(Box::new(DeclarationNode {
            typ,
            names
        }))
    }

    fn parse_type(&mut self) -> Result<Box<TypeNode>, Error> {
        let specifier = self.parse_type_specifier()?;

        let width = if self.lookahead.kind == TokenKind::LeftBracket {
            self.match_token(TokenKind::LeftBracket)?;

            let width = self.parse_number()?;

            self.match_token(TokenKind::RightBracket)?;

            Some(width)
        } else {
            None
        };

        Ok(Box::new(TypeNode {
            specifier,
            width
        }))
    }

    fn parse_type_specifier(&mut self) -> Result<Box<TypeSpecifierNode>, Error> {
        match self.lookahead.kind {
            TokenKind::InKeyword => {
                self.match_token(TokenKind::InKeyword)?;

                Ok(Box::new(TypeSpecifierNode::In))
            },
            TokenKind::OutKeyword => {
                self.match_token(TokenKind::OutKeyword)?;

                Ok(Box::new(TypeSpecifierNode::Out))
            },
            TokenKind::BlockKeyword => {
                self.match_token(TokenKind::BlockKeyword)?;

                let name = self.parse_identifier()?;

                Ok(Box::new(TypeSpecifierNode::Block(name)))
            },
            kind => Err(ErrorKind::UnexpectedToken2(self.lookahead.location, "type".to_string(), kind).into())
        }
    }

    fn parse_identifier_list(&mut self) -> Result<Vec<Box<IdentifierNode>>, Error> {
        let mut identifiers = Vec::<Box<IdentifierNode>>::new();

        identifiers.push(self.parse_identifier()?);

        while self.lookahead.kind == TokenKind::Comma {
            self.match_token(TokenKind::Comma)?;

            identifiers.push(self.parse_identifier()?);
        }

        Ok(identifiers)
    }

    fn parse_behaviour_statements(&mut self) -> Result<Vec<Box<BehaviourStatementNode>>, Error> {
        let mut behaviour_statements = Vec::<Box<BehaviourStatementNode>>::new();

        while self.lookahead.kind == TokenKind::Identifier {
            let behaviour_statement = self.parse_behaviour_statement()?;

            behaviour_statements.push(behaviour_statement);
        }

        Ok(behaviour_statements)
    }

    fn parse_behaviour_statement(&mut self) -> Result<Box<BehaviourStatementNode>, Error> {
        let target = self.parse_behaviour_identifier()?;

        self.match_token(TokenKind::Equals)?;

        let source = self.parse_expression()?;

        self.match_token(TokenKind::Semicolon)?;

        Ok(Box::new(BehaviourStatementNode {
            target,
            source,
            expression_type: None
        }))
    }

    fn parse_behaviour_identifier(&mut self) -> Result<Box<BehaviourIdentifierNode>, Error> {
        let name = self.parse_identifier()?;

        let property = if self.lookahead.kind == TokenKind::Dot {
            self.match_token(TokenKind::Dot)?;

            Some(self.parse_identifier()?)
        } else {
            None
        };

        let subscript = if self.lookahead.kind == TokenKind::LeftBracket {
            Some(self.parse_subscript()?)
        } else {
            None
        };

        Ok(Box::new(BehaviourIdentifierNode {
            name,
            property,
            subscript
        }))
    }

    fn parse_subscript(&mut self) -> Result<Box<SubscriptNode>, Error> {
        self.match_token(TokenKind::LeftBracket)?;

        let first = self.parse_number()?;

        let upper;
        let lower;

        if self.lookahead.kind == TokenKind::Colon {
            self.match_token(TokenKind::Colon)?;

            lower = self.parse_number()?;
            upper = Some(first);
        } else {
            lower = first;
            upper = None;
        }

        self.match_token(TokenKind::RightBracket)?;

        Ok(Box::new(SubscriptNode {
            upper,
            lower,
            upper_index: None,
            lower_index: None,
        }))
    }

    fn parse_expression(&mut self) -> Result<Box<ExpressionNode>, Error> {
        self.parse_bitwise_or_expression()
    }

    fn parse_bitwise_or_expression(&mut self) -> Result<Box<ExpressionNode>, Error> {
        let mut left = self.parse_bitwise_xor_expression()?;

        while self.lookahead.kind == TokenKind::OR {
            self.match_token(TokenKind::OR)?;

            let right = self.parse_bitwise_xor_expression()?;

            left = Box::new(ExpressionNode {
                data: ExpressionNodeData::Binary(BinaryOp::OR, left, right),
                typ: None,
            })
        }

        Ok(left)
    }

    fn parse_bitwise_xor_expression(&mut self) -> Result<Box<ExpressionNode>, Error> {
        let mut left = self.parse_bitwise_and_expression()?;

        while self.lookahead.kind == TokenKind::XOR {
            self.match_token(TokenKind::XOR)?;

            let right = self.parse_bitwise_and_expression()?;

            left = Box::new(ExpressionNode {
                data: ExpressionNodeData::Binary(BinaryOp::XOR, left, right),
                typ: None,
            })
        }

        Ok(left)
    }

    fn parse_bitwise_and_expression(&mut self) -> Result<Box<ExpressionNode>, Error> {
        let mut left = self.parse_unary_expression()?;

        while self.lookahead.kind == TokenKind::AND {
            self.match_token(TokenKind::AND)?;

            let right = self.parse_unary_expression()?;

            left = Box::new(ExpressionNode {
                data: ExpressionNodeData::Binary(BinaryOp::AND, left, right),
                typ: None,
            })
        }

        Ok(left)
    }

    fn parse_unary_expression(&mut self) -> Result<Box<ExpressionNode>, Error> {
        match self.lookahead.kind {
            TokenKind::NOT => {
                self.match_token(TokenKind::NOT)?;

                let operand = self.parse_unary_expression()?;

                Ok(Box::new(ExpressionNode {
                    data: ExpressionNodeData::Unary(UnaryOp::NOT, operand),
                    typ: None,
                }))
            },
            _ => self.parse_primary_expression()
        }
    }

    fn parse_primary_expression(&mut self) -> Result<Box<ExpressionNode>, Error> {
        match self.lookahead.kind {
            TokenKind::LeftParenthesis => {
                self.match_token(TokenKind::LeftParenthesis)?;

                let subexpression = self.parse_expression()?;

                self.match_token(TokenKind::RightParenthesis)?;

                Ok(subexpression)
            },
            TokenKind::Identifier => {
                let identifier = self.parse_behaviour_identifier()?;

                Ok(Box::new(ExpressionNode {
                    data: ExpressionNodeData::Variable(identifier),
                    typ: None,
                }))
            },
            TokenKind::Number => {
                let number = self.parse_number()?;

                Ok(Box::new(ExpressionNode {
                    data: ExpressionNodeData::Const(number),
                    typ: None,
                }))
            },
            kind => Err(ErrorKind::UnexpectedToken2(self.lookahead.location, "expression".to_string(), kind).into())
        }
    }
}
